File: examples/_go/disk.go
package main

import (
	"github.com/grafana/cog/generated/go/common"
	"github.com/grafana/cog/generated/go/dashboard"
	"github.com/grafana/cog/generated/go/table"
	"github.com/grafana/cog/generated/go/timeseries"
)

func diskIOTimeseries() *timeseries.PanelBuilder {
	return defaultTimeseries().
		Title("Disk I/O").
		FillOpacity(0).
		Unit("Bps").
		WithTarget(
			basicPrometheusQuery(`rate(node_disk_read_bytes_total{job="integrations/raspberrypi-node", instance="$instance", device!=""}[$__rate_interval])`, "{{device}} read"),
		).
		WithTarget(
			basicPrometheusQuery(`rate(node_disk_written_bytes_total{job="integrations/raspberrypi-node", instance="$instance", device!=""}[$__rate_interval])`, "{{device}} written"),
		).
		WithTarget(
			basicPrometheusQuery(`rate(node_disk_io_time_seconds_total{job="integrations/raspberrypi-node", instance="$instance", device!=""}[$__rate_interval])`, "{{device}} IO time"),
		).
		// Overrides configuration
		OverrideByRegexp("/ io time/", []dashboard.DynamicConfigValue{
			{Id: "unit", Value: "percentunit"},
		})
}

func diskSpaceUsageTable() *table.PanelBuilder {
	return table.NewPanelBuilder().
		Title("Disk Space Usage").
		Align(common.FieldTextAlignmentAuto).
		Unit("decbytes").
		CellHeight(common.TableCellHeightSm).
		Footer(common.NewTableFooterOptionsBuilder().CountRows(false).Reducer([]string{"sum"})).
		WithTarget(
			tablePrometheusQuery(`max by (mountpoint) (node_filesystem_size_bytes{job="integrations/raspberrypi-node", instance="$instance", fstype!=""})`, "A"),
		).
		WithTarget(
			tablePrometheusQuery(`max by (mountpoint) (node_filesystem_avail_bytes{job="integrations/raspberrypi-node", instance="$instance", fstype!=""})`, "B"),
		).
		// Transformations
		WithTransformation(dashboard.DataTransformerConfig{
			Id: "groupBy",
			Options: map[string]any{
				"fields": map[string]any{
					"Value #A": map[string]any{
						"aggregations": []string{"lastNotNull"},
						"operation":    "aggregate",
					},
					"Value #B": map[string]any{
						"aggregations": []string{"lastNotNull"},
						"operation":    "aggregate",
					},
					"mountpoint": map[string]any{
						"aggregations": []string{},
						"operation":    "groupby",
					},
				},
			},
		}).
		WithTransformation(dashboard.DataTransformerConfig{
			Id:      "merge",
			Options: map[string]any{},
		}).
		WithTransformation(dashboard.DataTransformerConfig{
			Id: "calculateField",
			Options: map[string]any{
				"alias": "Used",
				"binary": map[string]any{
					"left":     "Value #A (lastNotNull)",
					"operator": "-",
					"reducer":  "sum",
					"right":    "Value #B (lastNotNull)",
				},
				"mode": "binary",
				"reduce": map[string]any{
					"reducer": "sum",
				},
			},
		}).
		WithTransformation(dashboard.DataTransformerConfig{
			Id: "calculateField",
			Options: map[string]any{
				"alias": "Used, %",
				"binary": map[string]any{
					"left":     "Used",
					"operator": "/",
					"reducer":  "sum",
					"right":    "Value #A (lastNotNull)",
				},
				"mode": "binary",
				"reduce": map[string]any{
					"reducer": "sum",
				},
			},
		}).
		WithTransformation(dashboard.DataTransformerConfig{
			Id: "organize",
			Options: map[string]any{
				"excludeByName": map[string]any{},
				"indexByName":   map[string]any{},
				"renameByName": map[string]any{
					"Value #A (lastNotNull)": "Size",
					"Value #B (lastNotNull)": "Available",
					"mountpoint":             "Mounted on",
				},
			},
		}).
		WithTransformation(dashboard.DataTransformerConfig{
			Id: "sortBy",
			Options: map[string]any{
				"fields": map[string]any{},
				"sort": []map[string]any{
					{"field": "Mounted on"},
				},
			},
		}).

		// Overrides configuration
		OverrideByName("Mounted on", []dashboard.DynamicConfigValue{
			{Id: "custom.width", Value: 260},
		}).
		OverrideByName("Size", []dashboard.DynamicConfigValue{
			{Id: "custom.width", Value: 93},
		}).
		OverrideByName("Used", []dashboard.DynamicConfigValue{
			{Id: "custom.width", Value: 72},
		}).
		OverrideByName("Available", []dashboard.DynamicConfigValue{
			{Id: "custom.width", Value: 88},
		}).
		OverrideByName("Used, %", []dashboard.DynamicConfigValue{
			{Id: "unit", Value: "percentunit"},
			{Id: "custom.cellOptions", Value: struct {
				Mode string `json:"mode"`
				Type string `json:"type"`
			}{
				Mode: "gradient",
				Type: "gauge",
			}},
			{Id: "min", Value: 0},
			{Id: "max", Value: 1},
		})
}

File: examples/typescript/disk.ts
import {PanelBuilder as TimeseriesPanelBuilder} from "../../generated/typescript/src/timeseries";
import {PanelBuilder as TablePanelBuilder} from "../../generated/typescript/src/table";
import {basicPrometheusQuery, defaultTimeseries, tablePrometheusQuery} from "./common";
import {
    FieldTextAlignment,
    TableCellDisplayMode,
    TableCellHeight,
    TableFooterOptionsBuilder
} from "../../generated/typescript/src/common";

export const diskIOTimeseries = (): TimeseriesPanelBuilder => {
    return defaultTimeseries()
        .title("Disk I/O")
        .fillOpacity(0)
        .unit("Bps")
        .withTarget(
            basicPrometheusQuery(`rate(node_disk_read_bytes_total{job="integrations/raspberrypi-node", instance="$instance", device!=""}[$__rate_interval])`, "{{device}} read"),
        )
        .withTarget(
            basicPrometheusQuery(`rate(node_disk_written_bytes_total{job="integrations/raspberrypi-node", instance="$instance", device!=""}[$__rate_interval])`, "{{device}} written"),
        )
        .withTarget(
            basicPrometheusQuery(`rate(node_disk_io_time_seconds_total{job="integrations/raspberrypi-node", instance="$instance", device!=""}[$__rate_interval])`, "{{device}} IO time"),
        )
        .overrideByRegexp("/ io time/", [
            {id: "unit", value: "percentunit"},
        ]);
};

export const diskSpaceUsageTable = (): TablePanelBuilder => {
    return new TablePanelBuilder()
        .title("Disk Space Usage")
        .align(FieldTextAlignment.Auto)
        .cellOptions({type: TableCellDisplayMode.Auto})
        .cellHeight(TableCellHeight.Sm)
        .footer(new TableFooterOptionsBuilder().countRows(false).reducer(["sum"]))
        .unit("decbytes")
        .withTarget(
            tablePrometheusQuery(`max by (mountpoint) (node_filesystem_size_bytes{job="integrations/raspberrypi-node", instance="$instance", fstype!=""})`, "A"),
        )
        .withTarget(
            tablePrometheusQuery(`max by (mountpoint) (node_filesystem_avail_bytes{job="integrations/raspberrypi-node", instance="$instance", fstype!=""})`, "B"),
        )

        // Transformations
        .withTransformation({
            id: "groupBy",
            options: {
                fields: {
                    "Value #A": {
                        aggregations: ["lastNotNull"],
                        operation: "aggregate"
                    },
                    "Value #B": {
                        aggregations: ["lastNotNull"],
                        operation: "aggregate"
                    },
                    mountpoint: {
                        aggregations: [],
                        operation: "groupby"
                    }
                }
            }
        })
        .withTransformation({
            id: "merge",
            options: {}
        })
        .withTransformation({
            id: "calculateField",
            options: {
                alias: "Used",
                binary: {
                    left: "Value #A (lastNotNull)",
                    operator: "-",
                    reducer: "sum",
                    right: "Value #B (lastNotNull)"
                },
                mode: "binary",
                reduce: {reducer: "sum"}
            }
        })
        .withTransformation({
            id: "calculateField",
            options: {
                alias: "Used, %",
                binary: {
                    left: "Used",
                    operator: "/",
                    reducer: "sum",
                    right: "Value #A (lastNotNull)"
                },
                mode: "binary",
                reduce: {"reducer": "sum"}
            }
        })
        .withTransformation({
            id: "organize",
            options: {
                excludeByName: {},
                indexByName: {},
                renameByName: {
                    "Value #A (lastNotNull)": "Size",
                    "Value #B (lastNotNull)": "Available",
                    mountpoint: "Mounted on"
                }
            }
        })
        .withTransformation({
            id: "sortBy",
            options: {
                fields: {},
                sort: [
                    {field: "Mounted on"}
                ]
            }
        })

        // Overrides configuration
        .overrideByName("Mounted on", [{id: "custom.width", value: 260}])
        .overrideByName("Size", [{id: "custom.width", value: 93}])
        .overrideByName("Used", [{id: "custom.width", value: 72}])
        .overrideByName("Available", [{id: "custom.width", value: 88}])
        .overrideByName("Used, %", [
            {id: "unit", value: "percentunit"},
            {
                id: "custom.cellOptions",
                value: {mode: "gradient", type: "gauge"}
            },
            {id: "max", value: 1},
            {id: "min", value: 0}
        ])
    ;
};
